sdk-base/src/main/java/com/mapbox/maps/plugin/Plugin.kt

```diff
@@ -1,42 +1,196 @@
 package com.mapbox.maps.plugin
 
+import com.mapbox.maps.plugin.animation.CameraAnimationsPlugin
+import com.mapbox.maps.plugin.annotation.AnnotationPlugin
+import com.mapbox.maps.plugin.attribution.AttributionPlugin
+import com.mapbox.maps.plugin.compass.CompassPlugin
+import com.mapbox.maps.plugin.gestures.GesturesPlugin
+import com.mapbox.maps.plugin.lifecycle.MapboxLifecyclePlugin
+import com.mapbox.maps.plugin.locationcomponent.LocationComponentPlugin
+import com.mapbox.maps.plugin.logo.LogoPlugin
+import com.mapbox.maps.plugin.overlay.MapOverlayPlugin
+import com.mapbox.maps.plugin.scalebar.ScaleBarPlugin
+
 /**
- * Class name for compass plugin.
- */
-const val PLUGIN_COMPASS_CLASS_NAME = "com.mapbox.maps.plugin.compass.CompassViewPlugin"
-/**
- * Class name for logo plugin.
- */
-const val PLUGIN_LOGO_CLASS_NAME = "com.mapbox.maps.plugin.logo.LogoViewPlugin"
-/**
- * Class name for gestures plugin.
- */
-const val PLUGIN_GESTURE_CLASS_NAME = "com.mapbox.maps.plugin.gestures.GesturesPluginImpl"
-/**
- * Class name for map attribution plugin.
- */
-const val PLUGIN_ATTRIBUTION_CLASS_NAME = "com.mapbox.maps.plugin.attribution.AttributionViewPlugin"
-/**
- * Class name for location component plugin.
- */
-const val PLUGIN_LOCATION_COMPONENT_CLASS_NAME = "com.mapbox.maps.plugin.locationcomponent.LocationComponentPluginImpl"
-/**
- * Class name for scale bar plugin.
- */
-const val PLUGIN_SCALE_BAR_CLASS_NAME = "com.mapbox.maps.plugin.scalebar.ScaleBarPluginImpl"
-/**
- * Class name for camera animations plugin.
- */
-const val PLUGIN_CAMERA_ANIMATIONS_CLASS_NAME = "com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl"
-/**
- * Class name for map overlay plugin.
- */
-const val PLUGIN_MAPOVERLAY_CLASS_NAME = "com.mapbox.maps.plugin.overlay.MapOverlayPluginImpl"
-/**
- * Class name for annotation plugin.
- */
-const val PLUGIN_ANNOTATION_CLASS_NAME = "com.mapbox.maps.plugin.annotation.AnnotationPluginImpl"
-/**
- * Class name for lifecycle plugin.
+ * Class describing plugin instance.
+ *
+ * For Mapbox plugins it is enough to provide plugin id from this class without plugin instance.
+ * For user-defined plugins both unique id and plugin instance must be provided.
  */
-const val PLUGIN_LIFE_CYCLE_CLASS_NAME = "com.mapbox.maps.plugin.lifecycle.MapboxLifecyclePluginImpl"
\ No newline at end of file
+sealed class Plugin(
+  /**
+   * Unique plugin id.
+   */
+  val id: String,
+  /**
+   * Plugin instance implementing [MapPlugin].
+   * Must be non-null for user-defined plugins
+   * or RuntimeException will occur when trying to initiate such plugin.
+   */
+  val instance: MapPlugin?
+) {
+  /**
+   * Class for Camera plugin responsible for map camera.
+   *
+   * @param pluginId unique id.
+   * @param pluginInstance instance of [CameraAnimationsPlugin]
+   */
+  class Camera(pluginId: String, pluginInstance: CameraAnimationsPlugin? = null) :
```

Comment:
Do we need to distinguish the different plugin types? Seems the type is not used since there's a different ID for each plugin. The Plugin class itself has everything we need for the plugin, no?

Reply:
And if we want to distinguish between custom plugin and Mapbox plugin, maybe we can separate it that way.

Reply:
It was my initial code but I've actually wanted to make API more typed if users would like to create implementations for our plugins. It also imho makes code clearer and gives us room to e.g. allow using one plugin of certain type (right now we are not limiting this as you asked 😉).

Reply:
Does it mean that we need to add a new plugin type to this class every time we come up with a new plugin?

Reply:
Yes, is it a problem? 😄 
Add corresponding class + Mapbox id for our default impl - sounds pretty OK to me

Reply:
I don't see an issue here, it's just adding more wrapping around things, and seems not actually used in the code.

I'm fine with current setup, since eventually the functionality we expose to users should work as designed.

Reply:
Gave it a final thought, will remove sealed class logic in favour of simplicity.